### ISSUE TITLE
Add optional storage type information to runtime configuration

### ISSUE THREAD

**Issue by ptitzler** [elyra-ai/elyra#1094]
Currently a user has to specify a cloud storage url/password/bucket in a runtime configuration, but not identify the type of storage. If we knew the storage type we could probably generate more user friendly links in the UI, that, for example point directly to a bucket instead of the storage root. 

Example from the pipeline run dialog:

![image](https://user-images.githubusercontent.com/13068832/99841980-e160cf80-2b23-11eb-931a-fafe0a979cf6.png)

Typically users will run the same pipeline multiple times resulting in "directory names" (key name prefixes in cos lingo) in the same bucket, each being very similar since they include date/time information:

![image](https://user-images.githubusercontent.com/13068832/99842136-24bb3e00-2b24-11eb-8673-d8c99a26a45f.png)

If the storage type is known we could generate more specific URLs, e.g. for minio:

`http://....com:31323/minio/pipeline-artifacts/untitled-1120112132/` instead of
`http://....com:31323/minio/pipeline-artifacts/`

**Comment by ptitzler** [comment]
I've come across two issues in the the past where our approach of treating every s3 provider  the same fails, meaning the object storage links the GUI exposes yield a 404:
- The GUI target URL might be `http://....com:31323/minio/pipeline-artifacts/untitled-1120112132/ ` for minio but `http://....com:31323/some/other/path/buckets/untitled-1120112132` for some other storage provider. This is where having type information available would make it possible to generate a valid URL.
- The S3 API endpoint is not the GUI endpoint (or there is no web gui enabled that can be used at all) This is the case in the Operate First environment on MOC.

**Comment by kevin-bates** [comment]
Hmm - I'm not sure what we'd do about the latter case - where the S3 endpoint is not the GUI endpoint.  I think you'd need another set of information entirely.

For the first example, would a `cos_path_prefix` value be more appropriate, rather than the provider type?  Or can we always infer the path prefix from the type?  (e.g., minio = `minio`, other-type = `some/other/path`)

In your example above, the bucket name for minio is `pipeline-artifacts`, correct?  So should the other example be `http://....com:31323/some/other/path/pipeline-artifacts/untitled-1120112132`  or does the other not reference the bucket name in the path at all?  Just trying to see if there meant to be some further synchronization between the two examples.

**Comment by ptitzler** [comment]
> Hmm - I'm not sure what we'd do about the latter case - where the S3 endpoint is not the GUI endpoint. I think you'd need another set of information entirely.

I agree

> For the first example, would a cos_path_prefix value be more appropriate, rather than the provider type? Or can we always infer the path prefix from the type? (e.g., minio = minio, other-type = some/other/path)

The latter statement (infer) should be always true, afaik, and we should be able to piece together the URL. I exploited that in the enhancement that exposes the location of the input artifacts in the KFP UI.

> In your example above, the bucket name for minio is pipeline-artifacts, correct? 

Yes

> So should the other example be http://....com:31323/some/other/path/pipeline-artifacts/untitled-1120112132 or does the other not reference the bucket name in the path at all? 

Yes, my apologies about the unintended ambiguity.  What I was trying to express was that given a _bucket name_ and a _path prefix_, the resulting GUI URL for minio is likely different from the GUI URL for IBM COS, aws, .... If the GUI URL pattern is understood (like we do for minio) adding a dedicated type that the user can choose from would make it possible to display user-friendly links for those providers. For others it would be a generic link that may or may not work.

**Comment by kevin-bates** [comment]
Thank you for the information.  I'd prefer to keep inference out of this if possible, so if we added something akin to `cos_path_prefix` then the specific URL could be composed as: `cos_endpoint` + `cos_path_prefix` + `cos_bucket` + `pipeline run name?` - does that sound right?

(What is the appropriate name for the last term?  I'm not sure if we're actually using ELYRA_RUN_NAME for that vs. just making that kind of thing available.)

**Comment by ptitzler** [comment]
> does that sound right?

Not necessarily. If you prefer we go the url-composition-logic-is-not-hard-coded route it might require a templating approach because there might be some static "strings" that are specific to one provider, e.g. for minio

```
{{endpoint}}/{{bucket}}/{{path_prefix}}/
```

vs other provider
```
{{endpoint}}/something-proprietary/{{bucket}}/{{path_prefix}}/
```

with:
- `endpoint`: the user-provided endpoint URL
- `bucket`: the user-provided bucket name
- `path_prefix`: the Elyra - generated unique pipeline id (same as today)

The combination of these would be the "base URL", which would be further augmented when Elyra stores run input and output artifacts as required, which brings me to your question

> (What is the appropriate name for the last term? I'm not sure if we're actually using ELYRA_RUN_NAME for that vs. just making that kind of thing available.)

Are you referring here to the proposal to store run output artifacts in a run-specific "directory"?

**Comment by kevin-bates** [comment]
Hmm, shouldn't your minio example really be:
```
{{endpoint}}/minio/{{bucket}}/{{path_prefix}}/
```
which is "something-proprietary" and arguably no different from another provider.

> Are you referring here to the proposal to store run output artifacts in a run-specific "sub-directory"?

Yes.  And your _question_ basically answers my question.  We're not using elyra-run-name ourselves, but making it available.  So the answer to what the last term should be called is something like `pipeline-id`.

Btw, it seems we're using "path prefix" in two different ways.  You appear to be using it to represent the `pipeline-id`, while I was using it to hold the "something proprietary" portion of the more specific COS URL.

I'm just trying to determine if the composition of the COS URL can be made in a deterministic manner,  and I think it can by treating the "filler" as a prefix rather than a type (from which the prefix must be inferred).  This probably means that the metadata meta-properties can't be an enum but more of a free-form string with "minio" the default value.

**Comment by ptitzler** [comment]
You are right, but I guess my example was a bit too specific. My concern is that we could encounter something like 
```
{{endpoint}}/something-proprietary/{{bucket}}/more-proprietary/{{path_prefix}}/
```

i just ran a quick test on IBM Cloud, where it turns out the GUI URL is a lot more complex because it includes additional metadata in various places:
```
https://cloud.ibm.com/objectstorage/crn%3Av1%3Abluemix%3Apublic%3Acloud-object-storage%3Aglobal%3Aa%2F90%3A?bucket=a-cross-region-de-in&bucketRegion=eu&endpoint=s3.eu.cloud-object-storage.appdomain.cloud&paneId=bucket_overview
```

For this to work a user would have to be able to specify a pattern.

> Btw, it seems we're using "path prefix" in two different ways. You appear to be using it to represent the pipeline-id, while I was using it to hold the "something proprietary" portion of the more specific COS URL.

Ah yes, I was using its S3 meaning, where it essentially defines a "directory path" where the object is stored. Like you are saying, today this path happens to be the unique pipeline id not something storage provider specific.

**Comment by kevin-bates** [comment]
Thank you for providing the IBM cloud example, because that throws a wrench into everything. :smile:

Since I think the "composition of pieces" will need to occur within code (baring a templated approach) and since we only support Minio-based COS today (at least that's my understanding), I'm wondering if we could make some initial improvement merely by applying some composition within our existing runtime code using the string 'minio' as the "cos_path_prefix", but hold off on introducing an additional metadata property (or two).  This would at least allow our users to access the pipeline directory directly.

Thoughts?

**Comment by ptitzler** [comment]
That'd be better than what we have today!

**Comment by kevin-bates** [comment]
After discussing this further in the scrum, we will revisit this for the 2.5 release when we plan to expose alternate storage capabilities and have a better idea of the best approach.
